refactor of sayAllHandler into speech

seanbudd opened the pull request:
### Link to issue number:

Follow up of #12228, which fixed #12225

### Summary of the issue:

`SpeechWithoutPauses` is only used by `sayAllHandler`, but the code lies in `speech\__init__.py`. Due to code changes `sayAllHandler` needs to instantiate a `SpeechWithoutPauses` instance, and would either introduce a circular dependency or require a singleton to be created when the instance is needed. 

### Description of how this pull request fixes the issue:

* `sayAllHandler` is moved to a new module - `speech.sayAll`. 
*  `SpeechWithoutPauses` is moved to it's own module
* `speech\__init__.py` has been moved to `speech\speech.py` so that  `speech.sayAll` can import the necessary functions from `speech`. 
* `sayAllHandler` top level functions have been refactor into a class. An instance of SayAllHandler is initialised when NVDA is started in a consistent manner with other initialisations in the code base. 

### Testing strategy:

Ensure #12225 is still fixed - ensure that sayAll still no longer mixes content from different runs.

Ensure the API changes match the changelog entry

### Known issues with pull request:

Linting has been disabled on `speech.py` and `speechWithoutPauses.py` to be reenabled (tracked by #12377)

### Change log entry:

Developer Changes

Changes to existing changelog items:

```diff
- - `speakTextInfo` will no longer send speech through `speakWithoutPauses` if reason is `SAYALL`, as `sayAllhandler` does this manually now. (#12150)
+ - `speakTextInfo` will no longer send speech through `speakWithoutPauses` if reason is `SAYALL`, as `SayAllHandler` does this manually now. (#12150)
```

```diff
- - `speech.speakWithoutPauses` has been removed - please use `speech.SpeechWithoutPauses(speakFunc=speech.speak).speakWithoutPauses` instead. (#12195)
- - `speech.re_last_pause` has been removed - please use `speech.SpeechWithoutPauses.re_last_pause` instead. (#12195)
+ - `speech.speakWithoutPauses` has been removed - please use `speech.speechWithoutPauses.SpeechWithoutPauses(speakFunc=speech.speak).speakWithoutPauses` instead. (#12195, #12251)
+ - `speech.re_last_pause` has been removed - please use `speech.speechWithoutPauses.SpeechWithoutPauses.re_last_pause` instead. (#12195, #12251)
```

New developer changes:

- `sayAllHandler` has been moved to `speech.sayAll` (#12251):
  - `speech.sayAll.SayAllHandler` exposes the functions `stop`, `isRunning`, `readObjects`, `readText`, `lastSayAllMode`.
  - `SayAllHandler.stop` also resets the `SayAllHandler` `SpeechWithoutPauses` instance.
  - `CURSOR_REVIEW` and `CURSOR_CARET` has been replaced with `CURSOR.REVIEW` and `CURSOR.CARET`.


### Code Review Checklist:

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

feerrenrut commented:
Typically, I would expect any new feature that required `speakWithoutPauses` would want it's own instance. The management of it is important, and it would be quite complicated to share that with `sayAllHandler`. That said, if extending `sayAllHandler` of course it likely makes sense to use that instance. I think this PR makes the ownership of that instance much more clear.

I think this change log message is fine.

lukaszgo1 commented:
Hi @seanbudd Thanks for this PR! I agree that code is much more readable now. Unfortunately in some rare cases with code from this PR NVDA can crash due to circular imports.
STR:

1. Download [Report passwords 1.0](https://github.com/nvdaes/reportPasswords/releases/download/1.0/reportPasswords-1.0.nvda-addon) or for that matter any add-on which imports `gui` in the installTasks module.
2. Install it and then try to uninstall.

With the code from this PR NVDA crashes after restart with the following log:
```
ERROR - addonHandler.Addon.completeRemove (11:53:58.056) - MainThread (5644):
task 'onUninstall' on addon 'reportPasswords' failed
Traceback (most recent call last):
  File "addonHandler\__init__.py", line 330, in completeRemove
    self.runInstallTask("onUninstall")
  File "addonHandler\__init__.py", line 466, in runInstallTask
    self._installTasksModule=self.loadModule('installTasks')
  File "addonHandler\__init__.py", line 445, in loadModule
    return loader.load_module(fullname)
  File "C:\Python38\lib\pkgutil.py", line 285, in load_module
    mod = imp.load_module(fullname, self.file, self.filename, self.etc)
  File "C:\Python38\lib\imp.py", line 234, in load_module
    return load_source(name, filename, file)
  File "C:\Python38\lib\imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 702, in _load
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "userConfig\addons\reportPasswords\installTasks.py", line 7, in <module>
    import gui
  File "gui\__init__.py", line 18, in <module>
    import ui
  File "ui.py", line 21, in <module>
    import speech
  File "speech\__init__.py", line 16, in <module>
    import api
  File "api.py", line 12, in <module>
    import review
  File "review.py", line 10, in <module>
    from NVDAObjects import NVDAObject, NVDAObjectTextInfo
  File "NVDAObjects\__init__.py", line 18, in <module>
    import eventHandler
  File "eventHandler.py", line 16, in <module>
    import treeInterceptorHandler
  File "treeInterceptorHandler.py", line 11, in <module>
    import documentBase
  File "documentBase.py", line 7, in <module>
    from scriptHandler import isScriptWaiting
  File "scriptHandler.py", line 13, in <module>
    import sayAllHandler
  File "sayAllHandler.py", line 30, in <module>
    _speechWithoutPausesInstance = speech.SpeechWithoutPauses(speakFunc=speech.speak)
AttributeError: partially initialized module 'speech' has no attribute 'SpeechWithoutPauses' (most likely due to a circular import)
DEBUG - addonHandler.Addon.completeRemove (11:53:58.096) - MainThread (5644):
removing addon reportPasswords from _disabledAddons/_blockedAddons
DEBUG - addonHandler._getAvailableAddonsFromPath (11:53:58.096) - MainThread (5644):
Listing add-ons from D:\my_repos\nvda\source\userConfig\addons
DEBUG - addonHandler._getAvailableAddonsFromPath (11:53:58.096) - MainThread (5644):
Loading add-on from D:\my_repos\nvda\source\userConfig\addons\charInfo
DEBUG - addonHandler._getAvailableAddonsFromPath (11:53:58.096) - MainThread (5644):
Found add-on charInfo - 1.4. Requires API: (2017, 3, 0). Last-tested API: (2019, 3, 0)
DEBUG - addonHandler._getAvailableAddonsFromPath (11:53:58.096) - MainThread (5644):
Loading add-on from D:\my_repos\nvda\source\userConfig\addons\columnsReview
DEBUG - addonHandler._getAvailableAddonsFromPath (11:53:58.096) - MainThread (5644):
Found add-on columnsReview - 3.0-20210115-dev. Requires API: (2017, 3, 0). Last-tested API: (2020, 4, 0)
DEBUG - core.main (11:53:58.096) - MainThread (5644):
Initializing appModule Handler
DEBUG - core.main (11:53:58.096) - MainThread (5644):
Initializing NVDAHelper
DEBUG - core.main (11:53:58.156) - MainThread (5644):
Initializing tones
DEBUG - core.main (11:53:58.166) - MainThread (5644):
Speech Dictionary processing
DEBUG - speechDictHandler.SpeechDict.load (11:53:58.166) - MainThread (5644):
Loading speech dictionary 'D:\my_repos\nvda\source\userConfig\speechDicts\default.dic'...
DEBUG - speechDictHandler.SpeechDict.load (11:53:58.166) - MainThread (5644):
file 'D:\my_repos\nvda\source\userConfig\speechDicts\default.dic' not found.
DEBUG - speechDictHandler.SpeechDict.load (11:53:58.166) - MainThread (5644):
Loading speech dictionary 'D:\my_repos\nvda\source\builtin.dic'...
DEBUG - speechDictHandler.SpeechDict.load (11:53:58.166) - MainThread (5644):
3 loaded records.
CRITICAL - __main__ (11:53:58.176) - MainThread (5644):
core failure
Traceback (most recent call last):
  File "nvda.pyw", line 264, in <module>
    core.main()
  File "core.py", line 303, in main
    import speech
  File "speech\__init__.py", line 16, in <module>
    import api
  File "api.py", line 12, in <module>
    import review
  File "review.py", line 10, in <module>
    from NVDAObjects import NVDAObject, NVDAObjectTextInfo
  File "NVDAObjects\__init__.py", line 18, in <module>
    import eventHandler
  File "eventHandler.py", line 16, in <module>
    import treeInterceptorHandler
  File "treeInterceptorHandler.py", line 11, in <module>
    import documentBase
  File "documentBase.py", line 7, in <module>
    from scriptHandler import isScriptWaiting
  File "scriptHandler.py", line 13, in <module>
    import sayAllHandler
  File "sayAllHandler.py", line 30, in <module>
    _speechWithoutPausesInstance = speech.SpeechWithoutPauses(speakFunc=speech.speak)
AttributeError: partially initialized module 'speech' has no attribute 'SpeechWithoutPauses' (most likely due to a circular import)

```

Could you please look into this?

seanbudd commented on source/sayAll.py:
```diff
@@ -1,273 +1,289 @@
+		reader.nextLine()
+
+
+SayAllHandler = _SayAllHandler()
+
```
should I make all the methods of `_SayAllHandler` static instead of creating an instance?

seanbudd commented on source/sayAll.py:
```diff
@@ -1,273 +1,289 @@
+	@property
+	def speechWithoutPausesInstance(self) -> 'speech.SpeechWithoutPauses':
+		if self._speechWithoutPausesInstance is None:
+			log.error('attempting to use sayAll without calling initialize first')
+			self.initializeSpeechWithoutPauses()  # TODO: or raise Error?
+			# raise AttributeError('call initializeSpeechWithoutPauses first before using sayAllHandler')
```
possibly better to raise an error here - which could require another note in the changelog about calling `initializeSpeechWithoutPauses` before otherwise using `SayAllHandler`

seanbudd commented:
a simple fix for this may be to import `sayAllHandler` before the chain that eventually reaches `speech` (eg in `ui` or `gui`)

lukaszgo1 commented:
Why you've decided not to convert activesayall to a class after all? 
According to my test importing `sayAllHandler` in `gui` or in `ui` does not fix this. The simplest fix would be to import `speech` and `ui` lazily in the `gui` module. Having said that this is not a first time we are trying to work around a crash caused by some add-on importing `gui` in the installTasks module which suggests to me that `gui` is in a dire need of a refactor. Perhaps we should focus on making `gui's` layout more  flexible and then come back to this PR?

feerrenrut commented:
We agree that restructuring code would eventually resolve these kinds of issues. However, we'd like to see if we can find a minimal fix in case we run into issues during the refactor.

feerrenrut commented on source/core.py:
```diff
@@ -303,6 +303,9 @@ def main():
 	speech.initialize()
+	from speech import sayAll
+	log.debug("Initializing sayAllHandler")
+	sayAll.SayAllHandler.initializeSpeechWithoutPauses()
```
I think if we are going this far, then rather than make SayAllHandler a singleton, just construct it in core (passing speech function in). Other module that need it can be initialised after this and have the SayAllHandler instance passed to them, or if that is difficult with there current design, they can fetch it (at function level, not module level) from core.

seanbudd replied:
would it make more sense to do this in `speech.initialize()` so that they can fetch it from speech, not core?

seanbudd replied:
See 01b8c5f07

seanbudd commented:
I tried to start clean at master (@ 0efeee4b7dbdd243a7b880cf08dcbef284b48896 ) to make reviewing this easier. The commits from `4ebb450` are based clean off master



```
>>>git diff  0efeee4b7dbdd243a7b880cf08dcbef284b48896 4ebb450

diff --git a/source/speech/__init__.py b/source/speech/speech.py
similarity index 100%
rename from source/speech/__init__.py
rename to source/speech/speech.py
```

seanbudd commented:
Fixing the lint as a result of file moves was quite difficult. Here was my process

## linting `source\speech\speech.py` due to the file move

in the nvda virtual environment run the following (see b198d8f):

```
pip install autopep8==1.5.2 
autopep8 --global-config tests\lint\flake8.ini source\speech\speech.py -i -a -a
```

1. convert spaces to tabs with IDE (VS code) (see b198d8f)
1. manually fix long comments (see 5532502)
1. manually fix hanging indents (see 5532502)
1. remove unused variables and check if they are used globally (see 986ce76)

## fixing the imports in `source\speech\__init__.py` so that compatibility doesn't break  (see db99485)

using git bash, generate 2 files which can be copy and pasted into `source\speech\__init__.py`:

```
# for from .speech import * in speech\__init__.py
# functions
sed -nE "s/^def ([A-z_-]+).*$/\t\1,/p" source/speech/speech.py > imports.txt
# other variables
sed -nE "s/^([A-z_-]+) = .*$/\t\1,/p" source/speech/speech.py >> imports.txt
# sort
sort imports.txt -o imports.txt

# for __all__ in speech\__init__.py
# functions
sed -nE 's/^def ([A-z_-]+).*$/\t"\1",/p' source/speech/speech.py > all.txt
# other variables
sed -nE 's/^([A-z_-]+) = .*$/\t"\1",/p' source/speech/speech.py >> all.txt
# sort
sort all.txt -o all.txt
```

manually include the functions that were being imported from .types and .priorities

feerrenrut commented:
If this is just moved code, perhaps we should consider not fixing the lint warnings. Instead we could temporarily disable lint as a failure on this branch (so that we get the benefits of the other tests). For a large scale formatting change it would be best if it did the whole code base at once and then we could ignore that revision explicitly.

seanbudd commented:
> If this is just moved code, perhaps we should consider not fixing the lint warnings. Instead we could temporarily disable lint as a failure on this branch (so that we get the benefits of the other tests). For a large scale formatting change it would be best if it did the whole code base at once and then we could ignore that revision explicitly.

I could move the work done for linting `source\speech\speech.py` into a new branch based off of this one? It should be relatively easy to merge in once this has been (with a blame.ignoreRevsFile).

seanbudd committed:
move speech/__init__.py to speech/speech.py and move sayAllHandler.py to speech/sayAll.py

seanbudd committed:
refactor of sayAllHandler into speech. (#12251)

SpeechWithoutPauses is only used by sayAllHandler, but the code lies in speech\__init__.py. Due to code changes sayAllHandler needs to instantiate a SpeechWithoutPauses instance, and would either introduce a circular dependency or require a singleton to be created when the instance is needed.

Description of how this pull request fixes the issue:
- sayAllHandler is moved to a new module - speech.sayAll.
- SpeechWithoutPauses is moved to it's own module
- speech\__init__.py has been moved to speech\speech.py so that speech.sayAll can import the necessary functions from speech.
- sayAllHandler top level functions have been refactor into a class. An instance of SayAllHandler is initialised when NVDA is started in a consistent manner with other initialisations in the code base.

seanbudd commented:
git sadly doesn't pick up on the file move once we reintroduce `speech/__init__.py` so I've redone these with separate commits. I've also added a #noqa to `speech/speech.py` to be fixed up with the linted PR

feerrenrut commented:
Yes, typically with moved code, I'd recommend cleaning up the commit history (git rebase -i) so that there is a "pure moved code" commit, followed by a commit to make the modifications required. This makes it easier to follow what is just moved, and what has also changed. Then we merge without squashing.

seanbudd commented:
@feerrenrut - I would prefer to wait until after review to do another rebase as it will be quite the headache to rebase again if there are any suggested changes as I have done this process a couple times already.

seanbudd commented:
will this backwards upgrade compatibility need to stay forever? I've marked it for 2022.1 deprecation but it would still be an issue when upgrading from 2020.1.

seanbudd commented:
@feerrenrut - I've done the rebase - I will squash the last 3 commits and any other suggested changes from review in another rebase.

feerrenrut reviewed (changes requested):
Currently commit e76fbd20651d9bfa37dbf296c89bec895bcd1411 changes every line of sayAll.py and speech.py.
I assume this is a linting change, but there are other changes mixed in. It makes it hard to see the important changes and do the review.

I'd prefer if the linting changes were done in a separate PR. I'd suggest disabling the lint checker in this branch, we can re-enable it when just prior to merging. You may have to manually run the lint checker and specify the commit range (to be post commits for moving files)

seanbudd commented:
@feerrenrut  - something must have gone wrong with the rebase - it was just some whitespace changes that my IDE included (I had something configured wrong at some point while working on this). I've fixed this up now with another rebase. 

Linting has been disabled for `speech/speech.py` to be reenabled with the linting fix PR.

feerrenrut commented on source/appModules/powerpnt.py:
```diff
@@ -407,18 +407,18 @@ def _get_firstChild(self):
 		"""Pushes focus to the newly selected object."""
-		if getattr(self,"_isHandlingSelectionChange",False):
-			# #3394: A COM event can cause this function to run within itself.
+		if getattr(self, "_isHandlingSelectionChange", False):
```
There are still some whitespace changes here that are distracting from the change.

seanbudd replied:
ah that was my editor changing LF to CRLF and then causing the lint to fail. I've fixed this now

feerrenrut commented on source/speech/sayAll.py:
```diff
@@ -15,92 +18,278 @@
+		yield seq
+
+
+class SpeechWithoutPauses:
```
Did you consider putting this in a separate module?

seanbudd replied:
I recall putting this in the same module being suggested as this is the only place it is used, I'm happy to move this if you think it should go somewhere else.

feerrenrut replied:
I think that gives us more room for future designs, and doesn't cost much now or in maintenance.

feerrenrut commented on source/speech/sayAll.py:
```diff
@@ -15,92 +18,278 @@
-	"""
-	return bool(_activeSayAll())
+class _SayAllHandler:
+	def __init__(self):
+		self.lastSayAllMode = None
+		self.speechWithoutPausesInstance = SpeechWithoutPauses(speakFunc=speak)
```
Rather than import speech and create a `SpeechWithoutPauses` instance, could you have it passed into the constructor.

seanbudd replied:
speech is still imported for aspects of `SpeechWithoutPauses`, but I have included this change

feerrenrut commented on source/core.py:
```diff
@@ -303,6 +303,9 @@ def main():
 	speech.initialize()
+	from speech import sayAll
+	log.debug("Initializing sayAllHandler")
+	sayAll.initalize(sayAll.SpeechWithoutPauses(speakFunc=speech.speak))
```
I think it would also be fine to construct the `SpeechWithoutPauses` inside of the initialize method.

seanbudd commented:
I've moved `SpeechWithoutPauses` to its own module as suggested, and moved the initialization.

I also noticed the type I was using for CURSOR values is not in python3.7, but these values should probably be an Enum anyway, so I have refactored them into one.

feerrenrut reviewed (commented):
I think something has gone wrong with the rebase. Commit dd83725dd939ce0ad95571c9c446ab814e845c15 doesn't make sense to me. 

This code has likely been moved, but isn't deleted in this diff.

seanbudd commented:
> I think something has gone wrong with the rebase. Commit [dd83725](https://github.com/nvaccess/nvda/commit/dd83725dd939ce0ad95571c9c446ab814e845c15) doesn't make sense to me.
> 
> This code has likely been moved, but isn't deleted in this diff.

It was an attempt at tracking the changes in both files a bit easier, as `speechWithoutPauses.py` is just a small segment of what used to be `speech\__init__.py`, while the rest moved to `speech\speech.py`.

feerrenrut commented:
In that commit functions like `getObjectSpeech` are created in a new file `source/speech/speechWithoutPauses.py` despite never having anything to do with speechWithoutPauses.
You might need to take another look at it, I'm not sure the result is what you intended.

seanbudd commented:
> In that commit functions like `getObjectSpeech` are created in a new file `source/speech/speechWithoutPauses.py` despite never having anything to do with speechWithoutPauses.
> You might need to take another look at it, I'm not sure the result is what you intended.

Yes, the intentions was to copy `speech\__init__.py` entirely over, so that the changes in `speech\speechWithoutPauses.py` in this commit are easier to track (just removing irrelevant code). It would be good to keep the entire blame history for the functions in `speech\speechWithoutPauses.py` but I don't think git can track `speech\__init__.py` being split into two files, and then recreated with other content. I don't mind changing this to just moving the required functionality over. https://github.com/nvaccess/nvda/pull/12251/commits/5d2e6c197bb6fea5702fb4c85cb0447a683a2768#diff-b01881d490e0d24548bdc4e134c2b6f0192d6f417abe4081fbd47b7d3c5422a1

seanbudd committed:
move SpeechWithoutPauses out of speech.py

feerrenrut reviewed (approved):
Looks good @seanbudd!

JulienCochuyt commented:
Seems to have caused #12251

Btw, sorry if this has been discussed before, but why did you write `speech.sayAll.SayAllHandler` with an initial capital "S" instead of `speech.sayAll.sayAllHandler`? (The usual convention being to capitalize types and not other attributes.)

Seeing the class named `_SayAllHandler` rings the bell to the common pattern `class _C` / `C = _C` / `c = C()` which allowed in Python 2 for easier injection of derived classes, but I guess that's out of point here.

seanbudd commented on source/speech/speech.py:
```diff
@@ -115,9 +105,8 @@ def cancelSpeech():
-	sayAllHandler.stop()
-	sayAllHandler.getSpeechWithoutPauses().reset()
+	from speech import sayAll
+	sayAll.SayAllHandler.stop()
```
see #12395

seanbudd commented on source/speech/sayAll.py:
```diff
@@ -15,55 +22,78 @@
-	reader = _ObjectsReader(obj)
-	_activeSayAll = weakref.ref(reader)
-	reader.next()
+	def stop(self):
+		'''
+		Stops any active objects reader and resets the SayAllHandler's SpeechWithoutPauses instance
+		'''
+		active = self._getActiveSayAll()
+		if active:
+			active.stop()
+			self.speechWithoutPausesInstance.reset()
+
```
See #12396
